### PR TITLE
Format logs before sending them to the main process

### PR DESCRIPTION
**Pull request by namanjain**
Bionic sends all the log records to the main process and lets the main
process log them. In this case, the log formatting happens in the main
process, which requires us to send the entire argument list of the
logs. The communication between processes requires every argument to be
pickleable using cloudpickle, which seems unreasonable to enforce.

To avoid breaking logging, we eagerly format the log message the way
logging does (check LogRecord.getMessage). Although the logging works,
this has a side effect of changing the template and arguments for the
log in case users intercepts the LogRecord in their custom handlers.

### CODE REVIEW

File: bionic/executor.py [hunk 1]
```diff
@@ -145,7 +145,28 @@ def __init__(self, queue):
         self._queue = queue
 
     def emit(self, record):
+        # TODO We send all log records to the main process for logging,
+        # even when logging is disabled for their log levels. This is
+        # expensive as we pay IPC cost for the log records and pay the
+        # cost of formating the log record, which can be costly for
```
Comment (jqmp) [inline bionic/executor.py:151]: nit: s/formating/formatting/

File: bionic/executor.py [hunk 2]
```diff
@@ -145,7 +145,28 @@ def __init__(self, queue):
...
+        # cases where users have a ton of disabled debug logs.
+        # We should investigate if we can either caching the log levels
```
Comment (jqmp) [inline bionic/executor.py:153]: s/caching/cache/

File: bionic/executor.py [hunk 3]
```diff
@@ -145,7 +145,28 @@ def __init__(self, queue):
...
+            # communication between processes requires every argument
+            # to be pickleable using cloudpickle. Users can log
+            # unpickleable objects like clients or handlers and expect
+            # the logger to work. To avoid breaking logging, we eagerly
+            # format the log message the way logging does
+            # (check LogRecord.getMessage). This has a side effect of
+            # changing the template and arguments for the log in case
+            # users intercepts the LogRecord in their custom handlers.
+            if len(record.args) > 0:
+                record.msg = record.msg % record.args
```
Comment (jqmp) [inline bionic/executor.py:168]: Should we make a copy of the record before mutating it? I'm wondering if there could be other handlers in this process that will also receive this record.

Comment (namanjain) [reply bionic/executor.py:169]: Oh right, I guess users can add handlers to a logger inside the entity functions or we can use this outside Loky subprocesses too.

**Comment by namanjain** [comment]
> Looks good except for my question about copying, plus the extra edge case you mentioned on Slack.

Copying the record just in case, handled the edge case I mentioned in Slack with tests, and added an entry in release notes. PTAL.

File: docs/release-notes.rst [hunk 4]
```diff
@@ -72,6 +72,12 @@ For each release, we list the following types of change (in this order):
    in the "stable" docs (corresponding to the last release) but will be visible in the
    "latest" docs (corresponding to the master branch).
 
+Bug Fixes
+---------
+
+- Parallel execution introduced in 0.8.0 had a bug in logging, where logging breaks
```
Comment (jqmp) [inline docs/release-notes.rst:78]: "introduced in 0.8.0" should probably be in parentheses (or bracketed by commas). The way it is now, it looks restrictive, as if there were multiple kinds of parallel execution.

File: docs/release-notes.rst [hunk 5]
```diff
@@ -72,6 +72,12 @@ For each release, we list the following types of change (in this order):
...
+  when at least one of the arguments to the log message is unpickleable.
```
Comment (jqmp) [inline docs/release-notes.rst:79]: We might want to expand on "logging breaks": this causes lost log messages and warnings, but not a crash, right?

Comment (namanjain) [reply docs/release-notes.rst:79]: That's correct. I'll take another stab at this.
